libcfnet/file_stream.c

```diff
@@ -0,0 +1,943 @@
+ * @return true on success, otherwise false
+ */
+static bool __SendMessage(
+    SSL *conn, const char *msg, size_t len, bool eof, bool err)
+{
+    assert(conn != NULL);
+    assert(msg != NULL || len == 0);
+    assert(len <= MESSAGE_SIZE);
+
+    /* Set message length */
+    uint16_t header = len << 4;
+
+    /* Set Error flag */
+    if (err)
+    {
+        header |= (1 << 0);
+    }
+
+    /* Set End-of-File flag */
+    if (eof)
+    {
+        header |= (1 << 1);
+    }
+
+    /* Send header */
+    header = htons(header);
+    int ret = TLSSend(conn, (char *) &header, HEADER_SIZE);
+    if (ret != HEADER_SIZE)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to send message header during file stream (%d != %d)",
+            ret,
+            HEADER_SIZE);
+        return false;
+    }
+
+    if (len > 0)
+    {
+        /* Send payload */
+        ret = TLSSend(conn, msg, len);
+        if (ret != (int) len)
+        {
+            Log(LOG_LEVEL_ERR,
+                "Failed to send message payload during file stream (%d != %zu)",
+                ret,
+                len);
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/**
+ * @brief Send a message using the file stream protocol
+ *
+ * @param conn The SSL connection object
+ * @param msg The message to send
+ * @param len The length of the message to send (must be less or equal to
+ *            MESSAGE_SIZE Bytes)
+ * @param eof Set to true if this is the last message in a transaction,
+ *            otherwise false
+ * @return true on success, otherwise false
+ */
+static inline bool SendMessage(
+    SSL *conn, const char *msg, size_t len, bool eof)
+{
+    assert(conn != NULL);
+    assert(msg != NULL || len == 0);
+
+    return __SendMessage(conn, msg, len, eof, false);
+}
+
+/**
+ * @brief Receive a message using the file stream protocol
+ *
+ * @param conn The SSL connection object
+ * @param msg The message receive buffer (must be MESSAGE_SIZE Bytes large)
+ * @param len The length of the reveived message
+ * @param eof Is set to true if this was the last message in the transaction
+ * @return true on success, otherwise false
+ *
+ * @note RecvMessage fails if the communication is broken or if we received an
+ *       error from the remote host. In both cases, we should not try to flush
+ *       the stream.
+ */
+static bool RecvMessage(SSL *conn, char *msg, size_t *len, bool *eof)
+{
+    assert(conn != NULL);
+    assert(msg != NULL);
+    assert(len != NULL);
+    assert(eof != NULL);
+
+    /* TLSRecv() expects a buffer this size */
+    char recv_buffer[CF_BUFSIZE];
+
+    /* Receive header */
+    uint16_t header;
+    int ret = TLSRecv(conn, recv_buffer, HEADER_SIZE);
+    if (ret != HEADER_SIZE)
+    {
+        Log(LOG_LEVEL_ERR,
+            "Failed to receive message header during file stream (%d != %d)",
+            ret,
+            HEADER_SIZE);
+        return false;
+    }
+    memcpy(&header, recv_buffer, HEADER_SIZE);
```

Comment:
You could use `&header` as the buffer and avoid this `memcpy()`.

Reply:
Yup, I also thought so. But `TLSRecv()` expects a buffer of size `CF_BUFSIZE` and that you never request more than `CF_BUFSIZE - 1` Bytes. And the reason why, is that it actually writes a terminating `\0` at the end of the message. I spent a long time trying to figure out why my program crashed with memory violations :stuck_out_tongue_winking_eye: There is a comment some lines above. But it probably deserves more explanation.